### PR TITLE
Add text replacement methods

### PR DESCRIPTION
**Pull request by pie-flavor**
These are very much needed, given the relative underpoweredness of `TextTemplate`s.

### CODE REVIEW

**Comment by ryantheleach** [comment]
@pie-flavor Is this still Work In Progress? if so reply with `~wip` to mark your issue as Work In Progress or reply with `~qa` to mark it as ready for Quality Assurance, if you wish it to be reviewed / tested by contributors / the community.

**Comment by pie-flavor** [comment]
-qa

**Comment by pie-flavor** [comment]
~qa

**Comment by Defman** [comment]
Is Text not meant to be imutable: Represents an immutable tree-structure of formatted (text) components. Each instance consists of content and a list of children texts appended after the content of this text. The content of the text is available through one of the subclasses.

**Comment by ryantheleach** [comment]
Sure, but it's a utility function for something that is inherently difficult to do yourself 100% correctly the first time. 

Strings are immutable too, doesn't stop them having replace, but it's not something you would want to (ab)use all the time.

My 2c.

**Comment by dualspiral** [comment]
I don't see a problem with it as long as the javadocs make it clear (in both the body and the returns statement) that the method returns a copy of the `Text` object with the specified replacements.

**Comment by pie-flavor** [comment]
@The-Defman I'm confused - where's the issue? The original text never gets modified.

@dualspiral I'll add that, though originally I never thought it could be an issue - anyone using Text.replace has inevitably used String.replace which behaves the same way.

File: src/main/java/org/spongepowered/api/text/Text.java [hunk 1]
```diff
@@ -342,6 +346,116 @@ public final Text toText() {
...
+     * <p>This method produces a copy with the specified replacements. The
+     * original is not modified.</p>
+     *
+     * @param oldValue The pattern to replace
+     * @param newValue The value to replace with
+     * @param lossy Whether lossy mode is enabled
+     * @return The resulting (copied) text
+     */
+    public Text replace(Pattern oldValue, Text newValue, boolean lossy) {
+        Text t = this.children.isEmpty() ? this : this.toBuilder().removeAll().append(this.children.stream()
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/Text.java]: Please give this a longer name - single character variables should be reserved for `for`.

File: src/main/java/org/spongepowered/api/text/Text.java [hunk 2]
```diff
@@ -342,6 +346,116 @@ public final Text toText() {
...
+     * 
+     * <p>This method produces a copy with the specified replacements. The
+     * original is not modified.</p>
+     *
+     * @param oldValue The pattern to replace
+     * @param newValue The value to replace with
+     * @param lossy Whether lossy mode is enabled
+     * @return The resulting (copied) text
+     */
+    public Text replace(Pattern oldValue, Text newValue, boolean lossy) {
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/text/Text.java:307]: Could you add a few comments to this method explaining what's going on? Will be nice for the next person who works on this.

**Review by kashike — CHANGES_REQUESTED** [review]

**Comment by randombyte-developer** [comment]
Can this be added to the next 7.x release or is that against semver?

**Comment by pie-flavor** [comment]
No API methods get broken, so 7.x should be a valid target.

**Comment by gabizou** [comment]
Needs to be updated since the text was refactored.

**Comment by Cybermaxke** [comment]
@gabizou This targets API 7

**Comment by ImMorpheus** [comment]
@Cybermaxke you need to change the target branch to stable-7 then

**Comment by pie-flavor** [comment]
Oh good, the git history got clonked. Is this worth pursuing in its current state? Will there be another API 7 release?

**Comment by ImMorpheus** [comment]
@pie-flavor Yes, please update and rebase

**Comment by pie-flavor** [comment]
Looks all good from here.

**Comment by XakepSDK** [comment]
Maybe `replace(Text, Text)`?

**Comment by pie-flavor** [comment]
Properly searching for a Text would require remembering the effective formatting of the text in addition to the Text's own formatting, and additionally parsing children, greatly complicating the algorithm. And there wouldn't really be a point in a Text argument that's only ever used via toPlain.

**Review by ImMorpheus — APPROVED** [review]
Working as intended

**Comment by Lignium** [comment]
You need to add these methods to API 8.

**Pull request merged**